File: app/models/univariate.rb
class Univariate

  def initialize(input_data)
    input_data_array = input_data.split(",")
    input_data_array = input_data_array.map {|int| int.to_f }
    @input_data = input_data_array
  end

  def get_basic_stats
    output = {}
    vector = @input_data.to_scale
    output[:n] = vector.size
    output[:min] = vector.min
    output[:first_quartile] = vector.percentil(25)
    output[:mean] = vector.mean
    output[:median] = vector.median
    output[:third_quartile] = vector.percentil(75)
    output[:max] =  vector.max
    output[:s] = vector.sd
    output[:sigma] = vector.standard_deviation_population
    output[:variance_sam] = vector.variance
    output[:variance_pop] = vector.variance_population
    output[:ss] = vector.ss
    output[:kurtosis] = vector.kurtosis
    output[:histogram_values] = histogram(@input_data).values
    output[:histogram_intervals] = histogram(@input_data).keys
    output
  end

  def histogram(input_data)
    n = input_data.size
    output = {}
    bins = [input_data.uniq.size, 12].min
    min = input_data.min
    max = input_data.max
    range = max - min
    width = range / bins.to_f
    (bins+1).times do |i|
      lower = min - (width/2.0) + width*i 
      upper = min + (width/2.0) + width*i
      freq = input_data.count{|num| num >= lower && num < upper } / n.to_f
      output[[lower, upper]] = (freq * 200).round
    end
    output
  end

end

File: app/assets/javascripts/univariate.js
$(document).ready(function(){  

  $(".leaderboard").hide()

    var draw = function(hist_values, hist_interval){
      console.log(hist_values)
      var min = hist_interval[0]
      var max = hist_interval[0]

      for(i = 0; i < hist_interval.length; i++){
        if(hist_interval[i] < min){
          min = hist_interval[i]
        }
        if (hist_interval[i]>max){
          max = hist_interval[i]
        }
      }

      var max_height = hist_values[0]

      for(i = 0; i < hist_values.length; i++){
        if (hist_values[i] > max_height){
          max_height = hist_values[i]
        }
      }      

      var scale = 300 / max_height

      var dataset = hist_values;
      var range = max - min;
      var tick_width = range / 7.0;
      var axis = [];
      for(i=0 ;i<8 ;i++){
        axis.push((min + i * tick_width).toFixed(1) )
      }

      var w = 300;
      var h = 200;
      var barPadding = 1;
      var padding = 5;

      var svg = d3.select('.graph')
                    .append("svg")
                    .attr("fill", "teal")
                    .attr("width", w)
                    .attr("height", h);
 
      svg.selectAll("rect")
        .data(dataset)
        .enter()
        .append("rect")
        .attr("x", function(d,i){
          return i * (w / dataset.length)
        })
        .attr("y", function(d){
          return h - d * scale
        })
        .attr("width", w / dataset.length - barPadding)
        .attr("height", function(d){
          return d * scale;
        })

      var svg2 = d3.select('.graph')
                    .append("svg")
                    .attr("width", 300)
                    .attr("height", 25);

      svg2.selectAll("text")                   
        .data(axis)
        .enter()
        .append("text")
        .text(function(d){
          return d
        })
        .attr("x", function(d, i){
          return i *  (265 / 7) ;
        })
        .attr("y", 25);  
    }


  $("#compute-button").on('click', function(){

    var input_data = $("#textarea").val()

      $.ajax({
      url: '/univariate/compute',
      type: 'POST',
      data: { input_data: input_data },
      success: function(data){
        $("#n").html(data.n)
        $("#min").html(data.min)
        $("#first_qart").html(data.first_quartile.toFixed(3))
        $("#mean").html(data.mean.toFixed(3))
        $("#median").html(data.median.toFixed(3))
        $("#third_qart").html(data.third_quartile.toFixed(3))
        $("#max").html(data.max)
        $("#s").html(data.s.toFixed(3))
        $("#sigma").html(data.sigma.toFixed(3))
        $("#samp_var").html(data.variance_sam.toFixed(3))
        $("#pop_var").html(data.variance_pop.toFixed(3))
        $("#ss").html(data.ss.toFixed(3))
        $("#kurt").html(data.kurtosis.toFixed(3))
        $(".graph").empty()
        draw(data.histogram_values, data.histogram_intervals)
      }
      })
      $(".leaderboard").slideDown();
    
    
  })

})
